transformer_engine/common/fused_attn/utils.cu

```diff
@@ -393,6 +412,60 @@ __global__ void cu_seqlens_padded_to_offsets(NVTE_QKV_Layout_Group layout_group,
   }
 }
 
+__global__ void cu_seqlens_padded_to_offsets(NVTE_QKV_Layout_Group layout_group, int64_t b,
+                                             int64_t h, int64_t hg, int64_t d_qk, int64_t d_v,
+                                             const int32_t *cu_seqlens_q_padded,
+                                             const int32_t *cu_seqlens_kv_padded,
+                                             DType offset_dtype, void *offsets_q, void *offsets_k,
+                                             void *offsets_v, void *offsets_o) {
+  if (offset_dtype == DType::kInt32) {
+    cu_seqlens_padded_to_offsets_impl<int32_t>(
+        layout_group, b, h, hg, d_qk, d_v, cu_seqlens_q_padded, cu_seqlens_kv_padded,
+        reinterpret_cast<int32_t *>(offsets_q), reinterpret_cast<int32_t *>(offsets_k),
+        reinterpret_cast<int32_t *>(offsets_v), reinterpret_cast<int32_t *>(offsets_o));
+  } else {
+    assert(offset_dtype == DType::kInt64 && "expect int64");
+    cu_seqlens_padded_to_offsets_impl<int64_t>(
+        layout_group, b, h, hg, d_qk, d_v, cu_seqlens_q_padded, cu_seqlens_kv_padded,
+        reinterpret_cast<int64_t *>(offsets_q), reinterpret_cast<int64_t *>(offsets_k),
+        reinterpret_cast<int64_t *>(offsets_v), reinterpret_cast<int64_t *>(offsets_o));
+  }
+}
+
+DType get_ragged_offset_dtype(NVTE_QKV_Layout_Group layout_group, int64_t num_attn_heads,
+                              int64_t num_gqa_groups, int64_t max_seqlen_q, int64_t max_seqlen_kv,
+                              int64_t head_dim_qk, int64_t head_dim_v) {
+  std::array<int64_t, 4> offsets_qkvo{};
+  switch (layout_group) {
+    case NVTE_QKV_Layout_Group::NVTE_HD_HD_HD:
+      offsets_qkvo[0] = num_attn_heads * head_dim_qk * max_seqlen_q;
+      offsets_qkvo[1] = num_gqa_groups * head_dim_qk * max_seqlen_kv;
+      offsets_qkvo[2] = num_gqa_groups * head_dim_v * max_seqlen_kv;
+      break;
+    case NVTE_QKV_Layout_Group::NVTE_3HD:
+    case NVTE_QKV_Layout_Group::NVTE_H3D:
+      offsets_qkvo[0] = 3 * num_attn_heads * head_dim_qk * max_seqlen_q;
+      offsets_qkvo[1] = offsets_qkvo[0];
+      offsets_qkvo[2] = offsets_qkvo[0];
+      break;
+    case NVTE_QKV_Layout_Group::NVTE_HD_2HD:
+    case NVTE_QKV_Layout_Group::NVTE_HD_H2D:
+      offsets_qkvo[0] = num_attn_heads * head_dim_qk * max_seqlen_q;
+      offsets_qkvo[1] = 2 * num_gqa_groups * head_dim_qk * max_seqlen_kv;
+      offsets_qkvo[2] = offsets_qkvo[1];
+      break;
+  }
+
+  offsets_qkvo[3] = num_attn_heads * head_dim_qk * max_seqlen_q;
+
+  size_t max_offset = *std::max_element(offsets_qkvo.begin(), offsets_qkvo.end());
+  if (max_offset > std::numeric_limits<int32_t>::max()) {
```

Comment:
I think we should set the required type to INT64 when cuDNN >= 9.5. Otherwise, if some batches require INT32 and others INT64, we'd have to recompile the graph between batches because the `.set_data_type()` attribute has changed.

Reply:
Wouldn't a change to the parameters passed to this function also indicate a change the dimensions of the tensors requiring a new graph to be generated?

Reply:
As discussed we will use 64-bit for 9.5+ so the calls to cuDNN will only check the version. We will keep this function to report to the user that their config requires 64 bit if not available and recommend upgrading.

Reply:
Good point - different parameters would cause a new graph to be generated too.